components/mailgun/sources/new-log-data/new-log-data.js

```diff
@@ -0,0 +1,81 @@
+const common = require("../common-webhook");
+const { mailgun } = common.props;
+
+module.exports = {
+  ...common,
+  key: "new-log-data",
+  name: "New log data",
+  description:
+    "Emit an event when new data is logged in Mailgun's Control Panel. Occurs for most actions within the associated Mailgun account.",
+  version: "0.0.1",
+  dedupe: "unique",
+  props: {
+    mailgun,
+    domain: { propDefinition: [mailgun, "domain"] },
+    baseRegion: { propDefinition: [mailgun, "baseRegion"] },
+    timer: { propDefinition: [mailgun, "timer"] },
+    db: "$.service.db",
+  },
+  hooks: {
+    async deploy() {
+      // Emits sample events on the first run during deploy.
+      let mailgunEvents = await this.mailgun.getMailgunEvents(
+        this.domain,
+        null,
+        10
+      );
+      if (mailgunEvents.items.length === 0) {
+        console.log(`No data available, skipping iteration`);
+        return;
+      }
+      mailgunEvents.items.forEach(this.emitEvent);
+      const last = new String(mailgunEvents.paging.last);
+      let idxSlash = last.lastIndexOf("/");
+      let lastUrlPart = last.substring(idxSlash + 1);
+      mailgunEvents = await this.mailgun.getMailgunEvents(
+        this.domain,
+        lastUrlPart,
+        10
+      );
+      this.db.set("next", mailgunEvents.paging.next);
+    },
+  },
+  methods: {
+    ...common.methods,
+    generateMeta(eventPayload) {
+      const ts = eventPayload.timestamp;
+      return {
+        id: eventPayload.id,
+        summary: `New data logged ${eventPayload.event}, with log-level: ${eventPayload["log-level"]}`,
+        ts,
+      };
+    },
+    emitEvent(eventPayload) {
+      const meta = this.generateMeta(eventPayload);
+      this.$emit(eventPayload, meta);
+    },
+  },
+  async run() {
+    let next = this.db.get("next");
+    let mailgunEvents = null;
+    do {
+      let nextUrlPart = "";
+      if (next) {
+        const idxSlash = next.lastIndexOf("/");
+        nextUrlPart = next.substring(idxSlash + 1);
+      }
+      mailgunEvents = await this.mailgun.getMailgunEvents(
+        this.domain,
+        nextUrlPart,
+        10
+      );
+      if (mailgunEvents.items.length === 0) {
+        console.log(`No data available, skipping iteration`);
+        break;
+      }
```

Comment:
Is checking that there are no items the right way to do this? I was looking at this part of the docs:
https://documentation.mailgun.com/en/latest/api-events.html#event-polling
It recommends checking the timestamp of the last event on the result page to see if the results are legit. Then if they're not, wait 15 seconds and retry.

Reply:
I don't think checking the timestamp is necessary because that algorithm applies to polling events taking into account a time range. So my thinking is that the timestamp check is to make sure results are trustworthy in that they have data older than the most recent time of the range.

In the documentation, right before the link you provide, it says that when the `last` page is reached and no items are returned, eventually the `next` that was included in said `last` page will do have items, once new events arrive.

> If the range is ascending then events will continue to be recorded but will not show in the current request time range pages that are provided. So after the most recent events have been retrieved and an empty result page has been reached, then requesting next page URL returned with the last page some time later will return events that occurred since then. And this can go on indefinitely.

Initially, when the source is activated, we query Mailgun from 1 days old to current time events, we then skip all the results and go to the last page, so my assumption is that we didn't have to check for any previous events and simply wait for new events to appear in the `next` page that we saved in the source's `db` prop.

Not sure if there are other thoughts on this? @michelle0927 @dylburger @jverce

Reply:
Thoughts?

Reply:
If what you have now is working fine, that's good with me!

Reply:
Ok, I left it as is.